java-checks/src/main/java/org/sonar/java/checks/spring/SpringIncompatibleTransactionalCheck.java

```diff
@@ -0,0 +1,196 @@
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S2229")
+public class SpringIncompatibleTransactionalCheck extends IssuableSubscriptionVisitor {
+
+  private static final String SPRING_TRANSACTIONAL_ANNOTATION = "org.springframework.transaction.annotation.Transactional";
+  private static final String JAVAX_TRANSACTIONAL_ANNOTATION = "javax.transaction.Transactional";
+
+  private static final String MANDATORY = "MANDATORY";
+  private static final String NESTED = "NESTED";
+  private static final String NEVER = "NEVER";
+  private static final String NOT_SUPPORTED = "NOT_SUPPORTED";
+  private static final String REQUIRED = "REQUIRED";
+  private static final String REQUIRES_NEW = "REQUIRES_NEW";
+  private static final String SUPPORTS = "SUPPORTS";
+
+  private static final Map<String, Set<String>> INCOMPATIBLE_PROPAGATION_MAP = buildIncompatiblePropagationMap();
+
+  private static Map<String, Set<String>> buildIncompatiblePropagationMap() {
+    Map<String, Set<String>> map = new HashMap<>();
+    map.put(null, new HashSet<>(Arrays.asList(MANDATORY, NESTED, REQUIRED, REQUIRES_NEW)));
+    map.put(MANDATORY, new HashSet<>(Arrays.asList(NESTED, NEVER, NOT_SUPPORTED, REQUIRES_NEW)));
+    map.put(NESTED, new HashSet<>(Arrays.asList(NESTED, NEVER, NOT_SUPPORTED, REQUIRES_NEW)));
+    map.put(NEVER, new HashSet<>(Arrays.asList(MANDATORY, NESTED, REQUIRED, REQUIRES_NEW)));
+    map.put(NOT_SUPPORTED, new HashSet<>(Arrays.asList(MANDATORY, NESTED, REQUIRED, REQUIRES_NEW)));
+    map.put(REQUIRED, new HashSet<>(Arrays.asList(NESTED, NEVER, NOT_SUPPORTED, REQUIRES_NEW)));
+    map.put(REQUIRES_NEW, new HashSet<>(Arrays.asList(NESTED, NEVER, NOT_SUPPORTED, REQUIRES_NEW)));
+    map.put(SUPPORTS, new HashSet<>(Arrays.asList(MANDATORY, NESTED, NEVER, NOT_SUPPORTED, REQUIRED, REQUIRES_NEW)));
+    return map;
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    ClassTree classTree = (ClassTree) tree;
+    if (!hasMethodsAnnotatedWithTransactional(classTree)) {
+      return;
+    }
+    Map<Symbol, String> methodsPropagationMap = collectMethodsPropagation(classTree);
+    if (hasSameValues(methodsPropagationMap)) {
+      return;
+    }
+    methodsPropagationMap
+      .forEach((symbol, propagation) -> checkMethodInvocation((MethodTree) symbol.declaration(), propagation, methodsPropagationMap));
+  }
+
+  private void checkMethodInvocation(MethodTree method, @Nullable String callerPropagation, Map<Symbol, String> methodsPropagationMap) {
+    BlockTree methodBody = method.block();
+    if (methodBody == null) {
+      return;
+    }
+    methodBody.accept(new BaseTreeVisitor() {
+      @Override
+      public void visitMethodInvocation(MethodInvocationTree methodInvocation) {
+        super.visitMethodInvocation(methodInvocation);
```

Comment:
If i'm correct you have no testcase that requires this line (which should be here). In other words, if I remove that line, no tests will fail : please add a case (a method invocation with another method invocation as argument for instance).

Reply:
I added a test case